packages/preferences-controller/src/PreferencesController.test.ts

```diff
@@ -32,6 +39,189 @@ describe('PreferencesController', () => {
+
+    it('should update selected address to first identity if the selected address was removed', () => {
+      const messenger = getControllerMessenger();
+      const controller = setupPreferencesController({
+        options: {
+          state: {
+            identities: {
+              '0x00': { address: '0x00', name: 'Account 1' },
+              '0x01': { address: '0x01', name: 'Account 2' },
+              '0x02': { address: '0x02', name: 'Account 3' },
+            },
+            selectedAddress: '0x02',
+          },
+        },
+        messenger,
+      });
+
+      messenger.publish(
+        'KeyringController:stateChange',
+        {
+          ...getDefaultKeyringState(),
+          keyrings: [{ accounts: ['0x00'], type: 'CustomKeyring' }],
+        },
+        [],
+      );
+
+      expect(controller.state.selectedAddress).toBe('0x00');
+    });
+
+    it('should not update existing identities', () => {
+      const identitiesState = {
+        '0x00': { address: '0x00', importTime: 1, name: 'Account 1' },
+        '0x01': { address: '0x01', importTime: 2, name: 'Account 2' },
+        '0x02': { address: '0x02', importTime: 3, name: 'Account 3' },
+      };
+      const messenger = getControllerMessenger();
+      const controller = setupPreferencesController({
+        options: {
+          state: {
+            identities: cloneDeep(identitiesState),
+            selectedAddress: '0x00',
+          },
+        },
+        messenger,
+      });
+
+      messenger.publish(
+        'KeyringController:stateChange',
+        {
+          ...getDefaultKeyringState(),
+          keyrings: [
+            { accounts: ['0x00', '0x01', '0x02'], type: 'CustomKeyring' },
+          ],
+        },
+        [],
+      );
+
+      expect(controller.state.identities).toStrictEqual(identitiesState);
+    });
+
+    it('should not duplicate accounts present in multiple keyrings', () => {
+      const identitiesState = {
+        '0x00': { address: '0x00', importTime: 1, name: 'Account 1' },
+        '0x01': { address: '0x01', importTime: 2, name: 'Account 2' },
+        '0x02': { address: '0x02', importTime: 3, name: 'Account 3' },
+      };
+      const messenger = getControllerMessenger();
+      const controller = setupPreferencesController({
+        options: {
+          state: {
+            identities: cloneDeep(identitiesState),
+            selectedAddress: '0x00',
+          },
+        },
+        messenger,
+      });
+
+      messenger.publish(
+        'KeyringController:stateChange',
+        {
+          ...getDefaultKeyringState(),
+          keyrings: [
+            { accounts: ['0x00', '0x01', '0x02'], type: 'CustomKeyring' },
+            { accounts: ['0x00', '0x01', '0x02'], type: 'CustomKeyring' },
+          ],
+        },
+        [],
+      );
+
+      expect(controller.state.identities).toStrictEqual(identitiesState);
+    });
+
+    it('should not update selected address if it is still among identities', () => {
```

Comment:
This test seems to be for `updateIdentities`. Should it go in another `describe`?

Reply:
Good catch! No this belongs here, but this test was meant to trigger a state change, not call updateIdentities. These tests were copied and modified, but I forgot to modify this one I guess

Reply:
Fixed here: https://github.com/MetaMask/core/pull/3799/commits/bd4819d8ac2e0bf2d7de1502a7244d59a6f8fef3